<a href="https://colab.research.google.com/github/DanielRam94m/College_ECCI_UCR/blob/master/Laboratorio_02_Aprendizaje_Mecanico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Laboratorio 02 Aprendizaje Mecánico**
### Árboles de Decisión
#### Celeste Balladares, B70873
#### Daniel Ramírez, B45675
---

**Importar librerias**

In [ ]:
# importar librerias
import pandas as pd
import numpy as np

**Obtener datasets**

In [ ]:
df_mushrooms = pd.read_csv (r'/content/mushrooms.csv')
print(df_mushrooms.shape)

y_mushrooms = df_mushrooms['class']
#print('Esta es "y"\n', y_iris.head(), '\n\n')

x_mushrooms = df_mushrooms.loc[ : , df_mushrooms.columns != 'class']
#print('Esta es "x"\n', x_iris.head())

df_mushrooms.head()

(8124, 23)


class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [ ]:
#Importar csv y guardar en un dataframe
df_iris = pd.read_csv (r'/content/iris.data')
df_iris.columns = ["sepal_length", "sepal_width", "petal_length", "petal_width", "Target"]
print(df_iris.shape)
print(df_iris.head(), '\n\n')

y_iris = df_iris['Target']
#print('Esta es "y"\n', y_iris.head(), '\n\n')

x_iris = df_iris.loc[ : , df_iris.columns != 'Target']
#print('Esta es "x"\n', x_iris.head())

(149, 5)
   sepal_length  sepal_width  petal_length  petal_width       Target
0           4.9          3.0           1.4          0.2  Iris-setosa
1           4.7          3.2           1.3          0.2  Iris-setosa
2           4.6          3.1           1.5          0.2  Iris-setosa
3           5.0          3.6           1.4          0.2  Iris-setosa
4           5.4          3.9           1.7          0.4  Iris-setosa 




In [ ]:
df_titanic = pd.read_csv (r'/content/titanic.csv')
print(df_titanic.shape)
print(df_titanic.head(), '\n\n')

y_titanic = df_titanic['Survived']
#print('Esta es clase "y"\n', y_titanic.head(), '\n\n')

x_titanic = df_titanic.loc[ : , df_titanic.columns != 'Survived']
#print('Esta es "x"\n', x_titanic.head())

(1309, 12)
   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN   

**Step 1) Función Gini**

In [ ]:
def Gini(y):
  '''
  Función Gini(y) que recibe un objeto pandas.Series 'y'. Calcula el coeficiente de Gini para dicha serie. 
  '''
  if isinstance(y, pd.Series):
    p = y.value_counts()/y.shape[0]
    gini = 1-(np.sum(p**2))
    return(gini)
  else:
    print("Objeto debe ser un Pandas.Series")

In [ ]:
#Ejemplo de prueba de funcion Gini
Gini(df_titanic.Survived)

0.4699322607020445

**Step 2) Función Gini Split**

In [ ]:
def Gini_split(ys):
  '''
  Función que recibe un arreglo con diversos objetos pandas.Series y calcula el coeficiente de Ginisplit para esta división.
  '''
  nodo_padre = ys.iloc[:, 0] # Columna padre
  nodo_izq = ys.iloc[:, 1] # Columna hijo izquierdo
  nodo_der = ys.iloc[:, -1] # Columna hijo derecho
  nodo_padre = nodo_padre.dropna()
  nodo_izq = nodo_izq.dropna()
  nodo_der =nodo_der.dropna()
  num_izq = len(nodo_izq) / len(nodo_padre)
  num_der = len(nodo_der) / len(nodo_padre)
  gini_split = num_izq * Gini(nodo_izq) + num_der * Gini(nodo_der)
  informacion_ganada = Gini(nodo_padre) - gini_split 
  #print("info ganada",informacion_ganada)
  return informacion_ganada

In [ ]:
#Ejemplo de prueba de funcion Gini split
padrr = pd.Series([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1])
izqq = pd.Series([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1])
derr = pd.Series([0, 0, 0, 0, 1, 1, 1, 1])
ys = pd.concat([padrr, izqq, derr], axis=1)
#print(ys)
print('Informacion ganada', np.round(Gini_split(ys), 5))

Informacion ganada 0.11333


**Step 3) Metodo FIT**

### Clase Nodo

In [ ]:
class Nodo():
    def __init__(self, tipo=None, conteo=None, columna_split=None, valor_split=None, tipo_split=None, info_ganada=None, hijo_izq=None, hijo_der=None, clase =None):
        self.tipo = tipo
        self.conteo = conteo
        self.columna_split = columna_split
        self.valor_split = valor_split
        self.tipo_split = tipo_split
        self.hijo_izq = hijo_izq
        self.hijo_der = hijo_der
        self.info_ganada = info_ganada
        self.clase = clase

### Clase Árbol de Decisión

In [ ]:
from pandas.core.internals.base import DataManager
from operator import ilshift
class DecisionTree():
    def __init__(self):        
        # inicializa la raiz del árbol 
        self.root = None 
        # Condición de parada
        self.max_depth = None
        self.datos_split_min = 2
    
    def fit(self, x, y, max_depth = None):
        ''' función para entrenar el árbol'''
        self.max_depth = max_depth 
        data = pd.concat([x,y], axis=1)
        self.root = self.build_decision_tree(data)
    
    def build_decision_tree(self, data, my_depth =0):
        x = data.iloc[:,:-1]
        y = data.iloc[: , -1]
        num_datos = x.shape[0]
        if self.max_depth != None:
            if (my_depth < self.max_depth and num_datos >= self.datos_split_min):
                best_split = self.get_best_split(x, y)
                if best_split["gini"]>0:
                  # Recursion hijo izquierdo. Respuesta positiva
                  sub_arbol_izq = self.build_decision_tree(best_split["child-left"], my_depth+1)
                  # Recursion hijo derecho. Respuesta negativa
                  sub_arbol_der = self.build_decision_tree(best_split["child-right"], my_depth+1)
                  # retorna un nodo de decision
                  return Nodo(best_split["type"], best_split["count"], best_split["split-column"], 
                              best_split["split-value"], best_split["split-type"], best_split["gini"], sub_arbol_izq, sub_arbol_der) 
        else:
          if (num_datos >= self.datos_split_min):
            best_split = self.get_best_split(x, y)
            if best_split["gini"]>0:
              # Recursion hijo izquierdo. Respuesta positiva
              sub_arbol_izq = self.build_decision_tree(best_split["child-left"], my_depth+1)
              # Recursion hijo derecho. Respuesta negativa
              sub_arbol_der = self.build_decision_tree(best_split["child-right"], my_depth+1)
              # retorna un nodo de decision
              return Nodo(best_split["type"], best_split["count"], best_split["split-column"], 
                          best_split["split-value"], best_split["split-type"], best_split["gini"], sub_arbol_izq, sub_arbol_der)
 
        # Actualiza informacion nodo hoja
        clase_hoja = self.calcular_clase_hoja(y)
        # Retorna nodo hoja
        return Nodo(tipo='leaf', conteo=len(y), clase=clase_hoja)
    
    def get_best_split(self, x, y):
        #Lista para guardar el best split
        best_split = {}
        info_ganada_max = -float("inf")
        #Separamos los atributos numericos 
        numerical_atrib = x.select_dtypes(include=['int16', 'int32', 'int64', 'float16', 'float32', 'float64']).columns
        #Iteramos sobre cada posible atributo
        for atributo in x:
          atributos_valores = x.loc[:, atributo]
          if atributo in numerical_atrib:
            posibles_valores, atributo_valido = self.separacion_numerica(atributos_valores)
          else:
            posibles_valores, atributo_valido = self.separacion_categorica(atributos_valores)
          if(atributo_valido):
            # Iteramos sobre todos los posibles valores de un atributo
            for valor_split in posibles_valores:
                # get current split
                indices_hijo_izq, indices_hijo_der = self.split_nodes(x, atributo, valor_split, numerical_atrib)
                # Revisa que los nodos no se esten vacios
                if len(indices_hijo_izq)>0 and len(indices_hijo_der)>0:
                    y_df, y_izq, y_der = y, y.loc[indices_hijo_izq], y.loc[indices_hijo_der]
                    x_izq, x_der = x.loc[indices_hijo_izq], x.loc[indices_hijo_der]
                    ys = pd.concat([y_df, y_izq, y_der], axis=1)
                    # Se obtiene informacion ganada
                    info_ganada_act = Gini_split(ys)
                    # Actualiza el mejor split
                    if info_ganada_act > info_ganada_max:
                        #Se concatena los datos x con y
                        hijo_izq_data = pd.concat([x_izq, y_izq], axis=1)
                        hijo_der_data = pd.concat([x_der,y_der], axis=1)
                        # Actualiza informacion del mejor split
                        best_split["type"] = "split"
                        best_split["count"] = len(y_df)
                        best_split["split-column"] = atributo
                        best_split["split-value"] = valor_split
                        if atributo in numerical_atrib:
                          best_split["split-type"] = "numerical"
                        else:
                          best_split["split-type"] = "categorical"
                        best_split["gini"] = info_ganada_act
                        best_split["child-left"] = hijo_izq_data
                        best_split["child-right"] = hijo_der_data
                        info_ganada_max = info_ganada_act
                        
        # return best split
        return best_split
    
    def split_nodes(self, x, atributo, valor_split, numerical_atrib):
        ''' función que separa los atributos que se encuentran antes y después de la elegida ''' 
        if atributo in numerical_atrib:
          indices_izq = x.index[x[atributo] <= valor_split]
          indices_der = x.index[x[atributo] > valor_split]
        else:
          indices_izq = x.index[x[atributo] == valor_split]
          indices_der = x.index[x[atributo] !=  valor_split]
        return indices_izq, indices_der

    #3.a
    def calcular_clase_hoja(self, y):
      ''' funcion que calcula la clase del nodo hoja'''
      y = list(y)
      return max(y, key=y.count)    

    #3.c
    def separacion_numerica(self, atributo):
        max = atributo.max()
        min = atributo.min()
        if max!=min:
          pts_intermedios = np.linspace(min , max, num=12, retstep=True)
          pts_intermedios_array = pts_intermedios[0]
          pts_intermedios_array = pts_intermedios_array[1:11]
          return pts_intermedios_array, True
        else:
          return [], False

    #3.d
    def separacion_categorica(self, atributo):
        categories = atributo.unique()
        #Si tal atributo posee más de una categorís
        if 1 < len(categories):
            return categories, True
        else:
            return [], False

    def to_dict(self, node=None, espacio=" "):
        node_dic = {}
        if not node: 
          node = self.root
        if node.tipo == 'leaf':
          node_dic["type"] = node.tipo
          node_dic["class"] = node.clase
          node_dic["count"] = node.conteo
        else:
          node_dic["type"] = node.tipo
          node_dic["gini"] = node.info_ganada
          node_dic["count"] = node.conteo
          node_dic["split-type"] = node.tipo_split
          node_dic["split-column"] = node.columna_split
          node_dic["split-value"] = node.valor_split
          node_dic["child-left"] = self.to_dict(node.hijo_izq)
          node_dic["child-right"] = self.to_dict(node.hijo_der)
        return node_dic

    def predict(self, X):
      '''función que predice un nuevo dataset '''
      predicciones = []
      for num_row in range(len(X)):
        x = X.iloc[num_row]
        predicciones.append(self.make_prediction(x, self.root))
      print('\nPerdicciones hechas a cada entrada:\n\t', predicciones)
      return predicciones

    
    def make_prediction(self, x, nodo):
        '''función que predice fila por fila (entrada por entrada) '''
        if nodo.clase != None:
          return nodo.clase
        val_atributo = x[nodo.columna_split]
        #Si soy str voy con ==
        if isinstance(val_atributo, str):
          if val_atributo == nodo.valor_split:
              return self.make_prediction(x, nodo.hijo_izq)
          else:
              return self.make_prediction(x, nodo.hijo_der)
        #Si soy float voy con <=
        else:     
          if val_atributo <= nodo.valor_split:
              return self.make_prediction(x, nodo.hijo_izq)
          else:
              return self.make_prediction(x, nodo.hijo_der)



###Entrenando el árbol de decisión

In [ ]:
num_dataset = input('Elija el dataset que desea usar para el entrenamiento\n\t1) iris\n\t2) mushrooms\n\t3) titanic\n')
continua = False

if '1' == num_dataset:
  x_dataset = x_iris
  y_dataset = y_iris
  continua = True
elif '2' == num_dataset:
  x_dataset = x_mushrooms
  y_dataset = y_mushrooms
  continua = True
elif '3' == num_dataset:
  x_dataset = x_titanic
  y_dataset = y_titanic
  continua = True
else:
  print('\n\tValor no esperado, intente más tarde')

if continua == True:
  from sklearn.model_selection import train_test_split
  X_train, X_test, Y_train, Y_test = train_test_split(x_dataset, y_dataset, test_size=.2, random_state=41)

  #Testing train tree
  arbol_decision = DecisionTree()
  arbol_decision.fit(X_train,Y_train, max_depth=3)
  arbol = arbol_decision.to_dict()
  print(arbol)

Elija el dataset que desea usar para el entrenamiento
	1) iris
	2) mushrooms
	3) titanic
3
{'type': 'split', 'gini': 0.22224147079404197, 'count': 1047, 'split-type': 'categorical', 'split-column': 'Sex', 'split-value': 'male', 'child-left': {'type': 'split', 'gini': 0.04438689171821836, 'count': 667, 'split-type': 'numerical', 'split-column': 'Age', 'split-value': 14.889090909090909, 'child-left': {'type': 'split', 'gini': 0.2077566530691532, 'count': 48, 'split-type': 'numerical', 'split-column': 'SibSp', 'split-value': 2.1818181818181817, 'child-left': {'type': 'leaf', 'class': 1, 'count': 26}, 'child-right': {'type': 'leaf', 'class': 0, 'count': 22}}, 'child-right': {'type': 'split', 'gini': 0.012520773594727586, 'count': 473, 'split-type': 'numerical', 'split-column': 'Pclass', 'split-value': 1.1818181818181819, 'child-left': {'type': 'leaf', 'class': 0, 'count': 117}, 'child-right': {'type': 'leaf', 'class': 0, 'count': 356}}}, 'child-right': {'type': 'split', 'gini': 0.066679339

### Probando el árbol de decisión

In [ ]:
Y_pred = arbol_decision.predict(X_test) 
from sklearn.metrics import accuracy_score
accuracy_score(Y_test, Y_pred)


Perdicciones hechas a cada entrada:
	 [1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0]


0.851145038167939

### Matriz de confusión

In [ ]:
def calculate_confusion_matrix(predict, real):
    pclass = set(real.unique()).union(predict.unique())
    matrix = {}
    for real_index in pclass:
        valor = {}
        for predict_index in pclass:
            valor[predict_index] = 0
        matrix[real_index] = valor
    for predict_entry, real_entry in zip(predict, real):
        matrix[real_entry][predict_entry] = matrix[real_entry][predict_entry] + 1
    return matrix

#Prueba
conf_matrix = calculate_confusion_matrix(Y_test, pd.Series(Y_pred))
print('\nMi matriz de confusión:\n', conf_matrix)


Mi matriz de confusión:
 {'p': {'p': 777, 'e': 24}, 'e': {'p': 3, 'e': 821}}


###Conclusiones

En sus propias palabras: ¿qué mejoras o funcionalidades adicionales se le ocurre que podría incluir en su clase DecisionTree? Mencione al menos 2.
- Que el diccionario se pueda imprimir de forma más estructurada o entendible a la vista. :) 
-  Utilizamos estructuras de dataframes con pandas, sin embargo para el dataset de Titanic dura aproximadamente 1 min para entrenar el árbol, por lo que sería interesante probar si trabajar con estructuras numpy, pueda reducir el tiempo. 

Compare el rendimiento de su árbol con el de un árbol de la biblioteca sklearn, ¿qué tanto varían los resultados para el mismo set de datos de entrenamiento-prueba? 

Los porcentajes de predicción eran muy cercanos. Para el dataset Iris, nos dió 0.8667%, para Mushrooms 0.9834% y para Titanic 0.8511%.

